[ULP-1730] Added documentation for Send Phone Message Hook (#8805)
auth0/docs: articles/hooks/extensibility-points/send-phone-message.md

---
description: Learn how hooks can be used with the Send Phone Message extensibility point, which is available for MFA.

> beta: true

Reviewer: We should remove the 'beta' tag here

> beta: true

Author: ```suggestion
```

toc: true
topics:
    - hooks
    - extensibility-points
    - send-phone-message
    - custom-sms-gateway
    - sms 
contentType:
  - how-to
useCase: extensibility-hooks
v2: true
---

> # Send Phone Message

Reviewer: At the end we should add a 'Keep Reading' section that points to the Amazon example

Author: done

> At the Send Phone Message extensibility point, Hooks allow custom actions to be executed when a user is prompted for MFA. For example, you may use your preferred SMS gateway provider to get better pricing.

Reviewer: ```suggestion
At the Send Phone Message extensibility point, Hooks allow custom actions to be executed when a user is prompted for MFA. For example, you may use your preferred SMS gateway provider to get better pricing.
```
The Send Phone Message hook can be used to override how Auth0 sends SMS messages as part of the multi-factor authentication flows. Auth0 has built-in support for sending SMS messages with Twilio, but in some scenarios, you could want to use a different provider (e.g. you get better pricing in your region, or you already standardized with a different vendor).

Reviewer: I wanted to add a suggestion but it seemed I added a comment ;)

Author: ```suggestion
The Send Phone Message hook can be used to override how Auth0 sends SMS messages as part of the multi-factor authentication flows. Auth0 has built-in support for sending SMS messages with Twilio, but in some scenarios, you could want to use a different provider (e.g. you get better pricing in your region, or you already standardized with a different vendor).
```

The Send Phone Message extensibility point is available for [MFA](/multifactor-authentication) requests.

::: note
The `triggerId` for the Send Phone Message extensibility point is `send-phone-message`. To learn how to create Hooks for this extensibility point, see [Create New Hooks](/hooks/create).

> :::

Reviewer: This should be mentioned at the end of the document. How is the 'send-phone-message' triggerID relevant for customers ? If they want to create the hook using the Management API? If so, is there a document where we explain how to do that?

Author: It is explained on the linked page: https://auth0.com/docs/hooks/create . Or have I misunderstood?

Reviewer: Right, I missed the 'Management API' tab in the document. Let's move this to the bottom in a section with the title "Configuring the hook using the Management API"

Author: Got it, pushed this change.

@param {string} context.language - language used by login flow
@param {string} context.code - one time password
@param {string} context.ip - ip address
@param {string} context.user_agent - user agent making the authentication request
@param {object} context.client - object with details about the Auth0 application
@param {string} context.client.client_id - Auth0 application ID
@param {string} context.client.name - Auth0 application name
@param {object} context.client.client_metadata - metadata from client
@param {object} context.user - object representing the user
@param {string} context.user.user_id - Auth0 user ID
@param {string} context.user.name - user name
@param {string} context.user.email - user email
@param {object} context.user.app_metadata - metadata specific to user and application
@param {object} context.user.user_metadata - metadata specific to user
@param {function} cb - function (error, response)
*/
module.exports = function(recipient, text, context, cb) {
 // TODO: Add your code here 
 cb();
};
```

Please note:

::: note
The callback function (`cb`) at the end of the sample code signals completion and *must* be included.
:::

### Example parameters

```js
{

>   "recipient": "1-808-555-5555",

Reviewer 2: This usually comes with the `+`

Reviewer 3: I'll add this 😄

Reviewer 3: Do we need to add the "+" in order for this to work?

  "text": "Here is your one time password!",
  "context": {
    "message_type": "sms",
    "action": "enrollment",

>     "language": "english",

Reviewer 2: It will probably be a two character language code.

Reviewer 3: I'll update this so it says "en"

>     "code": "1234556ADSFA547865",

Reviewer 2: Usually six character code :)

Reviewer 3: I'll change this too :)

    "ip": "127.0.0.1",
    "user_agent": "Mozilla/5.0",
    "client": {
      "client_id": "1235",
      "name": "Test Application",
      "client_metadata": { }
    },
    "user": {
      "user_id": "auth0|test12345",

>       "name": "Billie Magnusson"

Reviewer 2: Missing `,`

Reviewer 3: 👍

  "context": {
    "message_type": "sms",
    "action": "enrollment",
    "language": "en",
    "code": "123456",
    "ip": "127.0.0.1",
    "user_agent": "Mozilla/5.0",
    "client": {
      "client_id": "1235",
      "name": "Test Application",
      "client_metadata": { }
    },
    "user": {
      "user_id": "auth0|test12345",
      "name": "Billie Magnusson",
      "email": "billie@email.com",
      "app_metadata": { },
      "user_metadata": { }
    }
  }
}
```

## Starter code response

Once you've customized the starter code, you can test the Hook using the Runner embedded in the Hook Editor. The Runner simulates a call to the Hook with the appropriate body and response. 

<%= include('../_includes/_test_runner_save_warning') %>

When you run a Hook based on the starter code, the response object is:

```
{
    "MessageID": "998a9ad1-c9b9-4b85-97b1-ac0305aa5532"
}
```

## Configuring the hook using the Management API

To configure this Hook using the Management API, please refer to the [Create New Hooks](/hooks/create) documentation.

::: note
The `triggerId` for the Send Phone Message extensibility point is `send-phone-message`.
:::

# Keep reading

> To view sample code on how to configure your Send Phone Message extensibility point using Amazon SNS, please follow the [Configuring a Custom SMS Gateway for MFA using Amazon SNS](/multifactor-authentication/send-phone-message-hook-configuration) guide. 

Author: @aaguiarz added the Keep Reading section as requested

    "user_agent": "Mozilla/5.0",
    "client": {
      "client_id": "1235",
      "name": "Test Application",
      "client_metadata": { }
    },
    "user": {
      "user_id": "auth0|test12345",
      "name": "Billie Magnusson",
      "email": "billie@email.com",
      "app_metadata": { },
      "user_metadata": { }
    }
  }
}
```

## Starter code response

Once you have customized the Hook code, you can test it using the Runner embedded in the Editor. The Runner simulates a call to the Hook with the appropriate body and response.

<%= include('../_includes/_test_runner_save_warning') %>

When you run a Hook based on the starter code, the response object is:

```
{
    "MessageID": "998a9ad1-c9b9-4b85-97b1-ac0305aa5532"
}
```

## Localization

The `context.language` parameter will always have one of the languages configured in the Tenant Settings (/universal-login/i18n). Depending on how you trigger the MFA flow, it will be calculated in different ways:

- If you use the [MFA API](/mfa/concepts/mfa-api) it will use the Accept-Language header from the request, and map it to a tenant language. If the language is not available, it will be set to the tenant default language.

> - If you use the New Universal Login Experience, it will use a combination of the Accept-Language header and the `ui_locales` parameter, as described [here](/i18n#language-selection).

Reviewer 4: ```suggestion
- If you use the New Universal Login Experience, it will use a combination of the Accept-Language header and the `ui_locales` parameter, as described in [Universal Login Internationalization](/universal-login/i18n#language-selection).
```

> ---

Author: This file will get linked from the Multi-Factor Authentication -> SMS -> Help text. Please see ULP-1734 for more info.

Reviewer 3: 👍

title: Send Phone Message 
description: Learn how hooks can be used with the Send Phone Message extensibility point, which is available for MFA.
beta: true
toc: true
topics:
    - hooks
    - extensibility-points
    - send-phone-message
    - custom-sms-gateway
    - sms 
contentType:
  - how-to
useCase: extensibility-hooks
v2: true
---

# Send Phone Message

> At the Send Phone Message extensibility point, Hooks allow custom actions to be executed when a user is prompted for MFA. For example, you may use your preferred SMS gateway provider to get better pricing, or you may make a voice call to the user with the MFA code instead of sending an SMS. 

Reviewer: ```suggestion
At the Send Phone Message extensibility point, Hooks allow custom actions to be executed when a user is prompted for MFA. For example, you may use your preferred SMS gateway provider to get better pricing.
```

@param {object} context - additional authorization context
@param {string} context.factor_type - 'first' or 'second'
@param {string} context.message_type - 'sms' or 'voice'
@param {string} context.action - 'enrollment' or 'authentication'
@param {string} context.language - language used by login flow
@param {string} context.code - one time password
@param {string} context.ip - ip address
@param {string} context.user_agent - user agent making the authentication request
@param {string} context.client_id - to send different messages depending on the client id
@param {string} context.name - to include it in the SMS message
@param {object} context.client_metadata - metadata from client
@param {object} context.user - To customize messages for the user
@param {function} cb - function (error, response)
*/
module.exports = function(recipient, text, context, cb) {
 // TODO: Add your code here 
 cb();
};
```

Please note:

::: note
The callback function (`cb`) at the end of the sample code signals completion and *must* be included.
:::

### Default response

When you run a Hook executed at the Send Phone Message extensibility point, the default response object is:

> 

Author: TODO: this needs to be added

Reviewer 3: 👍 Added

::: note
Hooks executed at the Send Phone Message extensibility point do not pass error messages to any Auth0 APIs.
:::

### Starter code response

Once you've customized the starter code, you can test the Hook using the Runner embedded in the Hook Editor. The Runner simulates a call to the Hook with the appropriate body and response. 

<%= include('../_includes/_test_runner_save_warning') %>

> When you run a Hook based on the starter code, the response object is:

Author: TODO: this needs to be added

Reviewer 3: removing the default response and added the response here 😄
